### PR TITLE
nixos/gnome3: enable uresourced

### LINKED ISSUE #15028: Dynamically change resource controls for user-.slice
This came up in discussions about setting resource guarantees for userspace processes. I'll update this ticket with more information later.

But in general, I believe we will want to reserve resources for `user-.slice` so that the users session can allocate them appropriately. However, we also need to ensure that an inactive session does not get such an allocation (inactive user e.g. the GDM greeter).

I suspect this logic should live in `systemd-logind`. Though I am not sure how the interaction with the usual resource control parameters should look like.

Relevant e-mails:
 * https://lore.kernel.org/linux-mm/4d3e00457bba40b25f3ac4fd376ba7306ffc4e68.camel@sipsolutions.net/ (initial mail floating the idea)
 * https://lore.kernel.org/linux-mm/20200305145554.GA5897@mtj.thefacebook.com/ (reply, should have most of the relevant information)
 * https://lore.kernel.org/linux-mm/CAOS58YM-HtmxwD7Q7g6ZzGsOJ_vWjHwb=7qpUwuZQEdeRrBifw@mail.gmail.com/ (further important notes from Tejun)

### PR DESCRIPTION
**Pull request by worldofpeace**
###### Motivation for this change
Uresourced was created to fix https://github.com/systemd/systemd/issues/15028, and is used in Fedora.

I've added a patch that has been sent upstream https://gitlab.freedesktop.org/benzea/uresourced/-/merge_requests/7

Currently need to test that it actually does anything with https://fedoraproject.org/wiki/Changes/Reserve_resources_for_active_user_WS#How_To_Test

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/system/uresourced.nix [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  options = {
+    services.uresourced = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+        Uresourced, a daemon that dynamically assigns a resource allocation to the active graphical user.
+        If the user has an active graphical session managed using systemd (e.g. GNOME), then the memory
+        allocation will be used to protect the sessions core processes (session.slice)";
+      '';
```
Comment (jtojnar) [inline nixos/modules/services/system/uresourced.nix:21]: indent

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 113913` at b317da80 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li><a href="https://gist.github.com/e6f443beea8b82f2394185d77af0db0e">uresourced</a></li></li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/os-specific/linux/uresourced/default.nix:24:4:
```
   |
24 |    ./0001-build-add-systemd-unit-dir-options.patch
   |    ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/os-specific/linux/uresourced/default.nix:25:4:
```
   |
25 |    ./fix-paths.patch
   |    ^
```
</li>
  </ul>
</details>

**Comment by worldofpeace** [comment]
Indented.

**Comment by rhendric** [comment]
Hey @worldofpeace, did you close this because you discovered it wasn't helpful, or some other reason? I had a notion of trying this myself.

**Comment by jtojnar** [comment]
See https://github.com/NixOS/nixpkgs/issues/121769

**Pull request closed without merging**